src/libraries/Native/Unix/System.Native/pal_signal.c

```diff
@@ -308,22 +511,87 @@ int32_t InitializeSignalHandlingCore()
-    // We don't handle ignored SIGINT/SIGQUIT signals. If we'd setup a handler, our child
-    // processes would reset to the default on exec causing them to terminate on these signals.
-    InstallSignalHandler(SIGINT , /* skipWhenSigIgn */ true);
-    InstallSignalHandler(SIGQUIT, /* skipWhenSigIgn */ true);
-    InstallSignalHandler(SIGCONT, /* skipWhenSigIgn */ false);
-    InstallSignalHandler(SIGCHLD, /* skipWhenSigIgn */ false);
-    InstallSignalHandler(SIGWINCH, /* skipWhenSigIgn */ false);
-
     return 1;
 }
 
+int32_t SystemNative_EnablePosixSignalHandling(int signalCode)
+{
+    assert(g_posixSignalHandler != NULL);
+    assert(signalCode > 0 && signalCode <= GetSignalMax());
+
+    bool installed;
+
+    pthread_mutex_lock(&lock);
+    {
+        installed = InstallSignalHandler(signalCode, SA_RESTART);
+
+        g_hasPosixSignalRegistrations[signalCode - 1] = installed;
+    }
+    pthread_mutex_unlock(&lock);
+
+    return installed ? 1 : 0;
+}
+
+void SystemNative_DisablePosixSignalHandling(int signalCode)
+{
+    assert(signalCode > 0 && signalCode <= GetSignalMax());
+
+    pthread_mutex_lock(&lock);
+    {
+        g_hasPosixSignalRegistrations[signalCode - 1] = false;
+
+        if (!(g_consoleTtouHandler && signalCode == SIGTTOU) &&
+            !(g_sigChldCallback && signalCode == SIGCHLD) &&
+            !(g_terminalInvalidationCallback && (signalCode == SIGCONT ||
+                                                 signalCode == SIGCHLD ||
+                                                 signalCode == SIGWINCH)))
+        {
+            RestoreSignalHandler(signalCode);
+        }
+    }
+    pthread_mutex_unlock(&lock);
+}
+
+void InstallTTOUHandlerForConsole(ConsoleSigTtouHandler handler)
+{
+    pthread_mutex_lock(&lock);
+    {
+        assert(g_consoleTtouHandler == NULL);
+        g_consoleTtouHandler = handler;
+
+        // When the process is running in background, changing terminal settings
+        // will stop it (default SIGTTOU action).
+        // We change SIGTTOU's disposition to get EINTR instead.
+        // This thread may be used to run a signal handler, which may write to
+        // stdout. We set SA_RESETHAND to avoid that handler's write loops infinitly
+        // on EINTR when the process is running in background and the terminal
+        // configured with TOSTOP.
+        RestoreSignalHandler(SIGTTOU);
+        InstallSignalHandler(SIGTTOU, (int)SA_RESETHAND);
+    }
+    pthread_mutex_unlock(&lock);
+}
+
+void UninstallTTOUHandlerForConsole(void)
+{
+    pthread_mutex_lock(&lock);
+    {
+        g_consoleTtouHandler = NULL;
+
+        RestoreSignalHandler(SIGTTOU);
+        if (g_hasPosixSignalRegistrations[SIGTTOU - 1])
+        {
+            InstallSignalHandler(SIGTTOU, SA_RESTART);
+        }
+    }
+    pthread_mutex_unlock(&lock);
+}
+
 #ifndef HAS_CONSOLE_SIGNALS
 
 int32_t SystemNative_InitializeTerminalAndSignalHandling()
 {
+    errno = ENOSYS;
```

Comment:
I've added this because the caller of this method should be checking `errno` when it returns zero.